### PR TITLE
Fixed the sample security voter to take into account the role hierarchy

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Doc fix? | yes |
| New docs? | no |
| Applies to | 2.3+ |
| Fixed tickets | #4389 |

I applied the solution suggested by @jdachtera because it seems to be the best one.

### CODE REVIEW

File: best_practices/security.rst [hunk 1]
```diff
@@ -147,6 +149,13 @@ the same ``getAuthorEmail`` logic you used above:
         const CREATE = 'create';
         const EDIT   = 'edit';
 
+        protected $roleHierarchy;
+
+        public function __construct(RoleHierarchyInterface $roleHierarchy)
+        {
+            $this->roleHierarchy = $roleHirarchy;
```
Comment (WouterJ) [inline best_practices/security.rst]: typo: `$roleHirarchy`

File: best_practices/security.rst [hunk 2]
```diff
@@ -173,6 +182,21 @@ the same ``getAuthorEmail`` logic you used above:
 
             return false;
         }
+
+        /**
+         * Checks if the user token has the given role taking into account the
+         * entire role hierarchy defined by the application.
+         */
+        protected function hasRole($roleName, TokenInterface $userToken)
```
Comment (WouterJ) [inline best_practices/security.rst]: wouldn't `userHasRole` be a better name. It now looks like a hasser/isser of the class itself

File: best_practices/security.rst [hunk 3]
```diff
@@ -147,6 +149,13 @@ the same ``getAuthorEmail`` logic you used above:
         const CREATE = 'create';
         const EDIT   = 'edit';
 
+        protected $roleHierarchy;
```
Comment (xabbuh) [inline best_practices/security.rst]: We should make this `private`.

File: best_practices/security.rst (hunk 2 above)
Comment (xabbuh) [inline best_practices/security.rst]: And it should be `private` imho.

**Comment by stof** [comment]
The RoleHierarchy service will not exist if you don't use the hierarchy in your security config

**Comment by javiereguiluz** [comment]
@stof thanks for the tip about the `security.role_hierarchy` service. I've just updated the sample code to take that into account.

File: best_practices/security.rst [hunk 4]
```diff
@@ -173,6 +182,25 @@ the same ``getAuthorEmail`` logic you used above:
...
         }
+
+        /**
+         * Checks if the user token has the given role taking into account the
+         * entire role hierarchy defined by the application.
+         */
+        private function userHasRole(TokenInterface $userToken, $roleName)
+        {
+            if (null === $this->roleHierarchy) {
+                return false;
```
Comment (stof) [inline best_practices/security.rst]: this is wrong. You should still use ``$userToken->getRoles()`` (which are the only reachable roles in this case).

File: best_practices/security.rst [hunk 5]
```diff
@@ -173,6 +182,25 @@ the same ``getAuthorEmail`` logic you used above:
 
             return false;
         }
+
+        /**
+         * Checks if the user token has the given role taking into account the
+         * entire role hierarchy defined by the application.
```
Comment (xabbuh) [inline best_practices/security.rst]: [...] hierarchy if defined [...] or [...] hierarchy as defined [...]?

File: best_practices/security.rst [hunk 6]
```diff
@@ -184,6 +212,9 @@ To enable the security voter in the application, define a new service:
         # ...
         post_voter:
             class:      AppBundle\Security\PostVoter
+            # beware that 'security.role_hierarchy' service is only defined
```
Comment (xabbuh) [inline best_practices/security.rst]: [...] that the 'security.role_hierarchy' [...]

File: best_practices/security.rst [hunk 7]
```diff
@@ -184,6 +212,9 @@ To enable the security voter in the application, define a new service:
...
+            # when the application's security config uses role hierarchy
```
Comment (xabbuh) [inline best_practices/security.rst]: [...] uses the role hierarchy

File: best_practices/security.rst [hunk 8]
```diff
@@ -173,6 +182,25 @@ the same ``getAuthorEmail`` logic you used above:
 
             return false;
         }
+
+        /**
+         * Checks if the user token has the given role taking into account the
+         * entire role hierarchy if defined by the application.
+         */
+        private function userHasRole(TokenInterface $userToken, $roleName)
```
Comment (aitboudad) [inline best_practices/security.rst]: I think you should pass UserInterface  here.
```
private function userHasRole(UserInterface $user, $roleName)
```

Comment (javiereguiluz) [inline best_practices/security.rst]: Much better. Fixed now. Thanks.

File: best_practices/security.rst [hunk 9]
```diff
@@ -173,6 +182,25 @@ the same ``getAuthorEmail`` logic you used above:
...
         }
+
+        /**
+         * Checks if the user token has the given role taking into account the
+         * entire role hierarchy if defined by the application.
+         */
+        private function userHasRole(UserInterface $user, $roleName)
+        {
+            if (null === $this->roleHierarchy) {
+                return in_array($roleName, $user->getRoles(), true);
```
Comment (stof) [inline best_practices/security.rst]: this is wrong. You should be using the roles from the token to be consistent with the way the RoleVoter is working

File: best_practices/security.rst [hunk 10]
```diff
@@ -173,6 +182,25 @@ the same ``getAuthorEmail`` logic you used above:
...
+            }
+
+            foreach ($this->roleHierarchy->getReachableRoles($user->getRoles()) as $role) {
```
Comment (stof) [inline best_practices/security.rst]: ``$user->getRoles()`` can represent roles as strings, which is not supported by ``getReachableRoles`` (the Token class is normalizing this, and ``getReachableRoles`` is meant to work with ``$token->getRoles()``

File: best_practices/security.rst (hunk 9 above)
Comment (aitboudad) [inline best_practices/security.rst]: there is no way to retrieve token https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php#L76

**Comment by javiereguiluz** [comment]
Closing it because we can't fix what we want to fix (because of the comments of @stof and @aitboudad)

**Pull request closed without merging**